hoverkite-protocol/src/lib.rs

```diff
@@ -194,20 +194,46 @@ pub enum Message {
     //     SecondaryCurrentPosition(SecondaryCurrentPosition),
 }
 
+#[cfg(feature = "std")]
+impl Message {
+    pub fn write_to_std(&self, writer: impl std::io::Write) -> std::io::Result<()> {
+        match self {
+            Self::Command(c) => c.write_to_std(writer),
+            Self::SecondaryCommand(sc) => sc.write_to_std(writer),
+        }
+    }
+}
+
+impl From<SecondaryCommand> for Message {
+    fn from(val: SecondaryCommand) -> Self {
+        Message::SecondaryCommand(val)
+    }
+}
+
+impl From<Command> for Message {
+    fn from(val: Command) -> Self {
+        Message::Command(val)
+    }
+}
+
 impl Message {
     pub fn parse(buf: &[u8]) -> nb::Result<Self, ParseError> {
-        let first = buf.get(0).ok_or(WouldBlock)?;
-        match first {
-            b'F' => {
-                let forward_length = buf.get(1).ok_or(WouldBlock)?;
-                if buf.len() < *forward_length as usize + 2 {
-                    return Err(WouldBlock);
+        match *buf {
+            [b'F', ref rest @ ..] => {
+                if let [forward_length, ref rest @ ..] = *rest {
+                    let forward_length = forward_length as usize;
+                    if rest.len() < forward_length {
+                        Err(WouldBlock)
+                    } else if rest.len() == forward_length {
+                        Ok(SecondaryCommand(Command::parse(rest)?).into())
```

Comment:
If `Command::parse` returns `Err(WouldBlock)`, then this should probably return a `ParseError` immediately, as it means that the `forward_length` doesn't match what it should be.